rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+* The graph element of both the matcher and relations work are relatively similar, but the denormalisation stages (the enricher and the merger) have less parallels. For example, the input to the merger is currently a group of work IDs to be merged with the stage having no knowledge of the graph representation, unlike the enricher which is likely to receive single works and needs access to the graph.
+
+Besides from the use of graphs and similarities to the matcher and merger, another aspect worth mentioning is the fact that we will need another store for denormalised works in addition to a store for merged works. Having two Elasticsearch indices coupled to corresponding ingestors is a possible solution (one for the current index of merged works and adding an additional one for denormalised works). There have been some previous discussions about replacing the recorder VHS with an Elasticsearch index, so it might be a worthwhile time to consider how this would work in this context, as here we would introducing an Elasticsearch index which was not the sink of the pipeline but used as an intermediate store.
+
+## Graph Store
+
+There has been some investigation into the use of graph databases such as Neo4j and AWS Neptune for storage and analysis of concepts in the pipeline. Using a specialised graph store might be a good general solution for relationships within our data.
+
+Neo4j for example would enable us, if desired, to store heterogeneous data types and relationships within a single graph:
+
+```
+[ Work A ] ── (matches with) ── [ Work C ]
+    │
+(parent of)
+    │
+[ Work B ]
+    │
+(relates to)
+    │
+    ├── [ Concept A ]
+    └── [ Concept B ]
+```
+
+Here square brackets indicate entities and parentheses indicate (directed) relationships.
+
+Having a single graph representation would enable richer queries and analytics of our data, and would likely be especially useful for machine learning work. However this may introduce complexity by breaking assumptions about data flow in the pipeline, such as if there were multiple stages which had the ability to write to the graph. This should therefore should be considered with caution. Whether as part of a single graph representation or a more specialised store solely for inter work relationships in the context of archives, using such a database to store relationships would aid implementation of the denormalisation stage, especially with the traversal and emitting of related works.
+
+Rather than storing the full work within the graph database, if only the IDs and relationships are stored this allows us more flexibility about where the graph analysis is performed in the pipeline and prevent data inconsistency errors in cases where the analysis was performed on pre-merged works. The current matcher database also stores solely the IDs, with the merger using these IDs to fetch the works themselves from the recorder VHS.
```

Comment:
Where would we store the work data then for retrieval?

e.g. I get a graph back, but need to send over the list of updated works, would I then look those IDs up from an index somewhere?

Would this mean extending `BigMessaging` to not only lookup from VHS< but potentially another index e.g. Elasticsearch?

Reply:
> Would this mean extending BigMessaging to not only lookup from VHS< but potentially another index e.g. Elasticsearch?

I do not understand why we would do that? I think we should work to remove the need for `BigMessaging` entirely.

Reply:
I think conflating things here, big messaging does not currently look up VHS (looks up S3). Definitely would not need to look up elasticsearch.

The application code would need to look up Elasticsearch to fetch the work data after receiving IDs from the graph store, as mentioned this is what we already do with the existing graph store implementation, where we look up the recorder VHS within the merger.

Reply:
My question is if we're passing IDs about and will need the whole works at stages in the pipeline, where does that lookup happen? It's not clear to me in this step.

Reply:
I don't think within this RFC I have suggested passing IDs around between services (which I don't think is a bad idea, just that not what this RFC focused on). The piece of text this comment refers to is about the graph store only containing IDs, but having a ES index to look up related works. When you look up the graph store in the enricher you will also need to look up the ES index, but that is within a single service (it will still receive works and emit works as messages as far as this RFC concerned).

Reply:
Super sorry if my question was unclear - but this answers it:

> the graph store only containing IDs, but having a ES index to look up related works

I do like the idea of passing around IDs in messaging, and having the receiving service do a lookup.